library: Add Language and Category filters

BharatAtbrat opened the pull request:
Initial implementation to add language and category filters for available demos.
Minor issue 
- ~~With the current implementation the `Dropdown` uses a `StringList` to store the names of the categories, and `Platform APIs` as a category only shows demos if the category is named `Platform`  in the string list. Therefore the category in the dropdown is also shown as `Platform` instead of the ideal `Platform APIs`~~
- Another thing I noticed is when the scroll is at the topmost position (default) it takes two clicks to select the overflowing demo categories. 
Screengrab of it in action:

https://github.com/user-attachments/assets/08eab5cb-3533-457a-b3e7-a4bb5b3d6edb


Let me know if there are any issues with the implementation :)

Current state [[commit](https://github.com/workbenchdev/Workbench/pull/967/commits/9bd062b96eae6cc544b46dbfac34e65558bdc3f3)]


https://github.com/user-attachments/assets/4dd3d5e1-267f-4506-b4f0-32b34acd5146

BharatAtbrat committed:
library: Add Language and Category filters
Initial implementation to add filters for demos

sonnyp committed:
Merge branch 'main' into atbrat/addfilters

sonnyp reviewed (commented):
Good start
I merged main into this branch, make sure to `git pull`


A few points

We shouldn't repeat the languages / categories - you can constructs the GTK from JavaScript

We shouldn't use human strings such as “All Languages” to keep track of the currently selected item. Use the index or some constant.

BharatAtbrat committed:
minor change: rename function

BharatAtbrat committed:
improvements:
Set dropdown to flat style
Construct dropdown values from JavaScript
Logic change for filtering, make use of index values instead of
strings

andyholmes commented on src/Library/Library.blp:
```diff
@@ -59,6 +59,18 @@ Adw.Window window {
+            }
+          }
+
           ListBox library_uncategorized {
```
The empty state here could probably use something:

![Screenshot of the library window, showing no results in search](https://github.com/user-attachments/assets/93f728d5-ca7d-4b8a-9a8c-93c33a72eaee)

If it were just one ListBox you'd use the placeholder feature, however the [HIG says](https://developer.gnome.org/hig/patterns/nav/search.html#search-results) *"Often a simple “No results” label is sufficient"*. Maybe one could be shown if there are no results?

BharatAtbrat replied:
That can be done yes

andyholmes replied:
That looks good, maybe add one of the [title style classes](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/main/style-classes.html#typography-styles) if one looks good, otherwise it can be tweaked it later :+1:

andyholmes reviewed (commented):
This is looking good.

I guess based on #510 this will eventually use a list model of demos with a GridView, so the filtering logic can be re-used later with GtkExpressions.

BharatAtbrat commented on src/Library/Library.js:
```diff
@@ -19,22 +21,51 @@ import { build } from "../../troll/src/builder.js";
   const demos = getDemos();
   const widgets_map = new Map();
   const category_map = new Map();
+  const language_model = new Gtk.StringList();
+  const category_model = new Gtk.StringList();
+  language_model.append("All Languages");
+  category_model.append("All Categories");
+  const language_check = ["All Languages"];
+  const category_check = ["All Categories"];
```
We have not discussed localization of these strings, but I agree that the category strings are not ideal. How do we localize them?

andyholmes replied:
I think the simplest thing to do is keep a static map, since there's a fixed amount of strings, like:

```js
let language_labels = {
  "javascript": _("JavaScript"),
  // etc
};
```

There seem to be strings for all languages in `Extensions.blp`, so the same strings can be reused without increasing translatables. It seems like the preference is to keep translations in `.blp` files, but I don't see an elegant way to keep them there with the `Library.js` parts being dynamic.

BharatAtbrat replied:
Okay, so I'll do that for both categories and languages. translatable category strings are present in the `Library.blp` except `_("Uncategorized");`
The linter does not like these changes 
```
  58:5   error  Replace `"javascript"` with `javascript`          prettier/prettier
  58:19  error  '_' is not defined                                no-undef
  59:5   error  Replace `"python"` with `python`                  prettier/prettier
  59:15  error  '_' is not defined                                no-undef
  60:5   error  Replace `"rust"` with `rust`                      prettier/prettier
```

andyholmes replied:
Adding this import should get the messages translated:

```js
import { gettext as _ } from "gettext";
```

I think the prettier style probably just wants the double-quotes removed from the dictionary keys, since they're not actually necessary.

BharatAtbrat committed:
minor addition: Add No results label on no match

BharatAtbrat committed:
improvements: Use translated strings for filters
Static maps for all category and Language filter strings
style change for "No results" label

andyholmes commented on src/Library/Library.js:
```diff
@@ -14,27 +16,78 @@ import { createSessionFromDemo } from "../sessions.js";
   const category_map = new Map();
+  const language_model = new Gtk.StringList();
+  const category_model = new Gtk.StringList();
+  language_model.append("All Languages");
+  category_model.append("All Categories");
+  const language_check = ["All Languages"];
+  const category_check = ["All Categories"];
```
Sorry, I missed these, which should also be localized.

After that, I think any UI tweaks can happen when we circle back with code search.

BharatAtbrat committed:
minor fix: localize All Category & Language string

andyholmes reviewed (approved):
LGTM!

sonnyp committed:
Merge branch 'main' into atbrat/addfilters

sonnyp committed:
fixup
